sdk/src/subnet_id.rs

```diff
@@ -1,311 +1,310 @@
+use fil_actors_runtime::cbor;
+use fnv::FnvHasher;
 use fvm_shared::address::Address;
 use lazy_static::lazy_static;
 use serde_tuple::{Deserialize_tuple, Serialize_tuple};
 use std::fmt;
-use std::path::{Path, PathBuf};
+use std::hash::{Hash, Hasher};
 use std::str::FromStr;
 
 use crate::error::Error;
-use crate::set_network_from_env;
 
+const SUBNET_ERR_TAG: &str = "subnetID";
+/// MaxChainID is the maximum chain ID value
+/// possible.
+const MAX_CHAIN_ID: u64 = 0xffffffffffffffff;
+
+/// SubnetID is a unique identifier for a subnet.
+/// It is composed of the chainID of the root network, and the address of
+/// all the subnet actors from the root to the corresponding level in the
+/// hierarchy where the subnet is spawned.
 #[derive(PartialEq, Eq, Hash, Clone, Debug, Serialize_tuple, Deserialize_tuple)]
 pub struct SubnetID {
-    parent: String,
-    actor: Address,
+    root: u64,
+    children: Vec<Address>,
 }
 
 lazy_static! {
-    pub static ref ROOTNET_ID: SubnetID = SubnetID {
-        parent: String::from("/root"),
-        actor: Address::new_id(0)
-    };
     pub static ref UNDEF: SubnetID = SubnetID {
-        parent: String::from("/"),
-        actor: Address::new_id(0)
+        root: 0,
+        children: vec![],
     };
 }
 
 impl SubnetID {
+    pub fn new(root_id: u64, children: Vec<Address>) -> Self {
+        Self {
+            root: root_id,
+            children,
+        }
+    }
+    /// Create a new subnet id from the root network id and the subnet actor
     pub fn new_from_parent(parent: &SubnetID, subnet_act: Address) -> Self {
+        let mut children = parent.children();
+        children.push(subnet_act);
         Self {
-            parent: parent.to_string(),
-            actor: subnet_act,
+            root: parent.root_id(),
+            children,
+        }
+    }
+
+    /// Creates a new rootnet SubnetID
+    pub fn new_root(root_id: u64) -> Self {
+        Self {
+            root: root_id,
+            children: vec![],
+        }
+    }
+
+    /// Returns true if the current subnet is the root network
+    pub fn is_root(&self) -> bool {
+        self.children_as_ref().is_empty()
+    }
+
+    /// Returns the chainID of the root network.
+    pub fn root_id(&self) -> u64 {
+        self.root
+    }
+
+    /// Returns the chainID of the current subnet
+    pub fn chain_id(&self) -> u64 {
+        if self.is_root() {
+            return self.root_id();
         }
+        let mut hasher = FnvHasher::default();
```

Comment:
Ah cool, I didn't know there was a hasher that output u64 directly. I'll switch to using this then.

Reply:
It would be good to include some examples in the form of tests. For example https://github.com/tjwebb/fnv-plus has a default bit length of 52, but also has named methods for 32 and 64. It would help users to want to produce the same chain ID on different platforms to include some documentation on what the Rust version uses.